### PR TITLE
Specify where send the funds during destroy

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1302

### CODE REVIEW

**Comment by roman-khimov** [comment]
1) What about other NEP-5 tokens?
2) Is there anything preventing user from transferring tokens to contract address after destroy?

**Comment by shargon** [comment]
> What about other NEP-5 tokens?

Is not possible to know the NEP-5 tokens that do you have without a list.

> Is there anything preventing user from transferring tokens to contract address after destroy?

No, non existing addresses it could be a valid address or a future contract.

**Comment by roman-khimov** [comment]
> Is not possible to know the NEP-5 tokens that do you have without a list.

That might be the reason for `RpcNep5Tracker` to be a standard function rather than plugin (at least for balance tracking, transfer tracking is much more expensive and less useful).

> No, non existing addresses it could be a valid address or a future contract.

I think it's worth blocking this address.

File: src/neo/SmartContract/InteropService.Contract.cs [hunk 1]
```diff
@@ -99,9 +100,30 @@ private static bool Contract_Update(ApplicationEngine engine)
...
                 ContractState contract = engine.Snapshot.Contracts.TryGet(hash);
                 if (contract == null) return true;
+
+                // Send funds
+
+                UInt160 sendToHash = new UInt160(sendTo.GetSpan());
+                if (engine.Snapshot.Contracts.TryGet(sendToHash)?.Payable == false) return false;
+
+                var balance = NativeContract.NEO.BalanceOf(engine.Snapshot, sendToHash);
+                if (balance > 0)
```
Comment (doubiliu) [inline src/neo/SmartContract/InteropService.Contract.cs:115]: Why check the balance of sendToHash? Does this mean that the amount of payee's account is 0 and no money can be received?

**Comment by erikzhang** [comment]
We can't stop someone from sending neo to a black hole address. So all we can do is reduce the chance of people making mistakes. Maybe this function should be done in the wallet or client, not in the core?

**Comment by doubiliu** [comment]
@erikzhang 
I think we can keep part of the inspection. When there is NEO or GAS in the current contract address to be deleted, it is forbidden to delete.Other inspection logic is handed over to the client

**Comment by erikzhang** [comment]
> When there is NEO or GAS in the current contract address to be deleted, it is forbidden to delete

It is a bit weird that the behaviour of a SYSCALL depends on the storage data of a contract. (Although it is a native contract.)

**Comment by shargon** [comment]
I think that the easier way is send the account as a parameter, you can send it to a black hole, but now it's not our problem.

**Comment by erikzhang** [comment]
The biggest problem here is that a SYSCALL will change the storage of a contract. This is unreasonable.

**Comment by shargon** [comment]
Only it's changed the neo/gas balance

**Comment by roman-khimov** [comment]
As a thought experiment --- can we somehow arrange to make proper transfer calls from `Contract.Destroy` to NEO/GAS/other NEP5 contracts? I guess the main problem is gonna be witness check, but we're sending money from the `CallingScriptHash` address, so probably it can work.

File: src/neo/SmartContract/InteropService.Contract.cs [hunk 2]
```diff
@@ -99,9 +100,30 @@ private static bool Contract_Update(ApplicationEngine engine)
 
             private static bool Contract_Destroy(ApplicationEngine engine)
             {
+                if (!engine.TryPop(out StackItem sendTo)) return false;
+
```
Comment (Tommo-L) [inline src/neo/SmartContract/InteropService.Contract.cs:104]: It's a little strange with a `sendTo` parameter. I think it's better to payback in contract, and we already have a NEP5 demo contract.

Comment (shargon) [reply src/neo/SmartContract/InteropService.Contract.cs:104]: I think that it's easier for developers, specify where they want to send the funds during destroy, ethereum has this behavior and I think that it's a good one, we have a deterministic address, ethereum not, and you can deploy it again, but if in the future the deploy cost it's higher like it was in neo2, i think that it's a big workaround that should not be required.

**Comment by erikzhang** [comment]
This function should be done in the wallet or client.

**Pull request closed without merging**